### PR TITLE
dbeaver/pro#3248 Put only one margin marker icon per line, add icons to annotation tooltip

### PR DESCRIPTION
**Pull request by E1izabeth**
image

```
ALTER TABLE Album 
	ADD CONSTRAINT dfk FOREIGN KEY (AlbumId) 
		REFERENCES Artdist (address) NOT VALID; 
```

### CODE REVIEW

File: plugins/org.jkiss.dbeaver.ui.editors.sql/src/org/jkiss/dbeaver/ui/editors/sql/util/SQLAnnotationHover.java [hunk 1]
```diff
@@ -285,5 +300,34 @@ public void setVisible(boolean visible) {
             super.setVisible(visible);
         }
     }
+    
+    private class AnnotationHyperlinkInfo {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.sql/src/org/jkiss/dbeaver/ui/editors/sql/util/SQLAnnotationHover.java:304]: Make it a `record` 🙂

File: plugins/org.jkiss.dbeaver.ui.editors.sql/src/org/jkiss/dbeaver/ui/editors/sql/semantics/SQLEditorSemanticMarkersManager.java [hunk 2]
```diff
@@ -132,25 +129,34 @@ private void updateMarkers() {
...
                         SQLSemanticErrorAnnotation annotation = new SQLSemanticErrorAnnotation(marker, problemInfo);
                         marker.setAttribute(SQLSemanticErrorAnnotation.MARKER_ATTRIBUTE_NAME, annotation);
-                        Position position = new Position(scriptItemPosition + problemInterval.a, problemInterval.length());
+                        Position position = new Position(problemOffset, problemInterval.length());
                         annotationModel.addAnnotation(annotation, position);
                         itemAnnotations.addLast(annotation);
-                    } catch (CoreException e) {
+                        severestAnnotationsByLine.compute(problemLine, (k, v) -> (v == null ||
+                            annotation.getProblemInfo().getSeverity().markerSeverity > v.getProblemInfo().getSeverity().markerSeverity
+                        ) ? annotation : v);
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.sql/src/org/jkiss/dbeaver/ui/editors/sql/semantics/SQLEditorSemanticMarkersManager.java]: This code is quite hard to read... :(
- You can use `Map#merge` that is identical to `Map#compute` but handles `null` keys automatically
- You can add getter for `getProblemInfo().getSeverity().markerSeverity` to shorten the code

File: plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/semantics/SQLQueryRecognitionProblemInfo.java [hunk 3]
```diff
@@ -97,6 +97,20 @@ public Interval getInterval() {
         return this.syntaxNode.getRealInterval();
     }
 
+    public String getExceptionMessage() {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/semantics/SQLQueryRecognitionProblemInfo.java:100]: I thought we already have such a method somewhere but I can't find it 🤔

File: plugins/org.jkiss.dbeaver.ui.editors.sql/src/org/jkiss/dbeaver/ui/editors/sql/util/SQLAnnotationHover.java (hunk 1 above)
Comment (E1izabeth) [reply plugins/org.jkiss.dbeaver.ui.editors.sql/src/org/jkiss/dbeaver/ui/editors/sql/util/SQLAnnotationHover.java:304]: Can't do that. I need editor field of an outer class.

**Review by ShadelessFox — APPROVED** [review]

**Comment by uslss** [comment]
verified

**Pull request merged**